File: node/src/domain/Pedido.ts
import CPF from "./CPF";
import CupomDesconto from "./CupomDesconto";
import Dimensao from "./Dimensao";
import Frete from "./Frete";
import ItemPedido from "./ItemPedido";
import Peso from "./Peso";
import Produto from "./Produto";

export default class Pedido {

    private id: string | undefined

    private cpf: CPF;

    private itens: ItemPedido[]

    private cupomDesconto: CupomDesconto

    private date: Date

    private frete: Frete

    constructor(cpf: string, date: Date) {
        this.cpf = new CPF(cpf);
        this.itens = []
        this.cupomDesconto = new CupomDesconto(0)
        this.date = date
        this.frete = new Frete()
        this.validate();
    }

    private validate() {
        if (!this.cpf.validate())
            throw new Error("Pedido não pode ser realizado, pq o CPF é invalido.")
    }

    public addItem(descricao: string, valor: number, quantidade: number, dimensao?: Dimensao, peso?: Peso) {
        this.frete.addProdutos(new Produto("", descricao, valor, dimensao, peso), quantidade)
        this.itens.push(new ItemPedido(descricao, valor, quantidade, dimensao, peso))
    }

    public addCupomDesconto(cupomDesconto: CupomDesconto) {
        this.cupomDesconto = cupomDesconto
    }

    public getTotal() {
        let total = 0
        this.itens.forEach((itemPedido) => {
            total += itemPedido.getTotal()
        })

        total += this.getValorFrete()

        if (this.cupomDesconto)
            return this.cupomDesconto.aplicarDesconto(total, this.date)


        return total
    }

    public getValorFrete() : number{
        return this.frete.getValor();
    }

    public gerarId(quantidadePedidos: number): string {
        if (!this.id) {
            this.id = new Date().getUTCFullYear().toString() + (quantidadePedidos + 1)
        }
        return this.id
    }

    public getIdPedido(): string | undefined{
        return this.id
    }
}

File: golang/src/domain/pedido.go
package domain

import (
	"errors"
	"time"
)

type Pedido struct {
	cpf   string
	itens []ItemPedido
	cupom Cupom
	Frete Frete
}

type ItemPedido struct {
	Item       Produto
	Quantidade int
}

type Cupom struct {
	Percentual    int
	DataExpiracao time.Time
}

func NewPedido(cpf string) (*Pedido, error) {
	ehValidoCPF := IsCPF(cpf)
	if !ehValidoCPF {
		return nil, errors.New("Pedido não pode ser realizado, pq o CPF é invalido.")
	}
	return &Pedido{cpf: cpf, itens: nil, cupom: Cupom{0, time.Now()}, Frete: Frete{}}, nil
}

func (pedido *Pedido) AddItem(descricao string, valor float64, quantidade int, dimensao Dimensao, peso Peso) {
	produto := Produto{
		Descricao: descricao,
		Valor:     valor,
		Dimensao:  dimensao,
		Peso:      peso,
	}

	itemPedido := ItemPedido{
		Item:       produto,
		Quantidade: quantidade,
	}

	pedido.itens = append(pedido.itens, itemPedido)
	pedido.Frete.AddProdutos(produto, quantidade)
}

func (pedido Pedido) GetTotal() float64 {
	total := 0.0
	for _, itemPedido := range pedido.itens {
		total += itemPedido.GetTotal()
	}
	total += pedido.Frete.GetValor()

	if pedido.cupom.Percentual > 0 {
		return pedido.cupom.AplicarDesconto(total)
	}
	return total
}

func (itemPedido ItemPedido) GetTotal() float64 {
	return itemPedido.Item.Valor * float64(itemPedido.Quantidade)
}

func (pedido *Pedido) AddCupomDesconto(cupom Cupom) {
	pedido.cupom = cupom
}

func (cupom Cupom) AplicarDesconto(valor float64) float64 {
	if cupom.CupomExpirado() {
		return valor
	}
	return valor - (valor * (float64(cupom.Percentual)) / 100.0)
}

func (cupom Cupom) CupomExpirado() bool {
	return time.Now().After(cupom.DataExpiracao)
}
